### PR TITLE
S3Express support

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category:
- New Feature

### Changelog entry:
Implemented support for S3Express buckets.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 421e39d105fceb14c92152df42a5d14f95258bcf with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59965/421e39d105fceb14c92152df42a5d14f95258bcf/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59965/421e39d105fceb14c92152df42a5d14f95258bcf/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by nickitat** [comment]
### S3 disk

1. Create a bucket of `Directory` type
2. Install correct bucket policy to grant all required permissions to your `S3` user (e.g. `"Action": "s3express:*"` to simply allow everything)
3. When configuring the storage policy please provide the `region` parameter

Storage configuration is the same as for ordinary `S3` and for example might look the following way:

``` sql
<storage_configuration>
    <disks>
        <s3_express>
            <type>s3</type>
            <endpoint>https://my-test-bucket--eun1-az1--x-s3.s3express-eun1-az1.eu-north-1.amazonaws.com/store/</endpoint>
            <region>eu-north-1</region>
            <access_key_id>...</access_key_id>
            <secret_access_key>...</secret_access_key>
        </s3_express>
    </disks>
    <policies>
        <s3_express>
            <volumes>
                <main>
                    <disk>s3_express</disk>
                </main>
            </volumes>
        </s3_express>
    </policies>
</storage_configuration>
```

And then create a table on the new storage:

``` sql
CREATE TABLE t
(
    a UInt64,
    s String
)
ENGINE = MergeTree
ORDER BY a
SETTINGS storage_policy = 's3_express';
```

### S3 storage

`S3` storage is also supported for but only for `Object URL` paths. Example:

``` sql
select * from s3('https://test-bucket--eun1-az1--x-s3.s3express-eun1-az1.eu-north-1.amazonaws.com/file.csv', ...)
```

it requires specifying bucket region in the config:

``` xml
<s3>
    <perf-bucket-url>
        <endpoint>https://test-bucket--eun1-az1--x-s3.s3express-eun1-az1.eu-north-1.amazonaws.com</endpoint>
        <region>eu-north-1</region>
    </perf-bucket-url>
</s3>
```

### Backups

It is also possible to store backups on the new tier:

``` sql
BACKUP TABLE t TO Disk('s3_express', 't.zip')

┌─id───────────────────────────────────┬─status─────────┐
│ c61f65ac-0d76-4390-8317-504a30ba7595 │ BACKUP_CREATED │
└──────────────────────────────────────┴────────────────┘
```

``` sql
RESTORE TABLE t AS t_restored FROM Disk('s3_express', 't.zip')

┌─id───────────────────────────────────┬─status───┐
│ 4870e829-8d76-4171-ae59-cffaf58dea04 │ RESTORED │
└──────────────────────────────────────┴──────────┘
```

**Comment by vitlibar** [comment]
> [x] Documentation is written (mandatory for new features)

What do you mean?

File: src/IO/S3/Requests.h [hunk 1]
```diff
@@ -21,12 +21,44 @@
...
+
+#include <base/defines.h>
 
 namespace DB::S3
 {
 
 namespace Model = Aws::S3::Model;
 
+/// Used only for S3Express
+namespace RequestChecksum
```
Comment (vitlibar) [inline src/IO/S3/Requests.h:36]: This namespace seems to be redundant and it seems it makes the code less clear.
These functions are very simple and it seems you can just do what you want directly without these functions.

File: src/IO/S3/Requests.h [hunk 2]
```diff
@@ -21,12 +21,44 @@
...
+    chassert(req.GetChecksumAlgorithm() == Aws::S3::Model::ChecksumAlgorithm::CRC32);
+    return Aws::Utils::HashingUtils::Base64Encode(Aws::Utils::HashingUtils::CalculateCRC32(*(req.GetBody())));
+}
+
+template <typename R>
+inline void setChecksumAlgorithm(R & request)
+{
+    if constexpr (requires { request.SetChecksumAlgorithm(Model::ChecksumAlgorithm::CRC32); })
+        request.SetChecksumAlgorithm(Model::ChecksumAlgorithm::CRC32);
+}
```
Comment (vitlibar) [inline src/IO/S3/Requests.h:59]: This template function looks like a bit of overcomplication. Which actual types of requests need this function `SetChecksumAlgorithm(Model::ChecksumAlgorithm::CRC32)` to be called? Only `UploadPartRequest`? Why can't you just call it immediately after calling `req.SetChecksumCRC32()` ?

**Comment by vitlibar** [comment]
BTW what is S3Express? And why it requires checksum to be sent?

**Comment by nickitat** [comment]
> > [x] Documentation is written (mandatory for new features)
> 
> What do you mean?

I mean the linked pr to docs: https://github.com/ClickHouse/clickhouse-docs/pull/2069

File: src/IO/S3/Requests.h (hunk 2 above)
Comment (nickitat) [reply src/IO/S3/Requests.h:59]: to my understanding it is needed for all requests with body. e.g. also `CreateMultipartUploadRequest`, `PutObjectRequest`.

**Comment by nickitat** [comment]
> BTW what is S3Express? And why it requires checksum to be sent?

https://aws.amazon.com/s3/storage-classes/express-one-zone/
idk why, but it is how the new tier works.

**Pull request merged**